### PR TITLE
[DataLoader] Make `batch` DataPipe sensitive to unbatch_level argument

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60840 [WIP] Example of DataPipes and DataFrames integration
* #60680 [DataLoader] Add notebook with DataPipes API example
* #60679 [DataLoader] Add Example Only `fork` DataPipe
* #60678 [DataLoader] Decorate ZipDataPipe as `zip`
* #60675 [DataLoader] Add simple `groupby` DataPipe
* **#60672 [DataLoader] Make `batch` DataPipe sensitive to unbatch_level argument**
* #60671 [DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument
* #60670 [DataLoader] Rename transform DataPipe to legacy_transform

Differential Revision: [D29461086](https://our.internmc.facebook.com/intern/diff/D29461086)

### CODE REVIEW

File: torch/utils/data/datapipes/iter/grouping.py [hunk 1]
```diff
@@ -43,16 +50,16 @@ def __iter__(self) -> Iterator[List[T_co]]:
...
             if not self.drop_last:
                 yield batch
-            batch.clear()
+            batch = []
 
     def __len__(self) -> int:
         if self.length is not None:
             return self.length
-        if isinstance(self.datapipe, Sized):
+        if isinstance(self.datapipe, Sized) and self.unbatch_level == 0:
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/grouping.py:62]: I don't think we need to check `self.unbatch_level == 0` since `unbatch` doesn't have `__len__` .
https://github.com/pytorch/pytorch/blob/fa030d1213600fa8626540f1b8857cbcbfeeb67c/torch/utils/data/datapipes/iter/grouping.py#L65

**Review by ejguan — APPROVED** [review]
LGTM except one nit

File: torch/utils/data/datapipes/iter/grouping.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/grouping.py:62]: This could work as optimization, imagine MP situation, when we need to send a message to request len of another pipe (and get special NotAvailable message back),
this shortcircuit will allow skipping this step.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461086).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461086).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461086).
